TermClose autocmd can free `curbuf`

- `nvim --version`: NVIM v0.4.0-1191-g740fb337d
- `vim -u DEFAULTS` (version: 8.1) behaves differently? It doesn't have a `TermClose` event
- Operating system/version: Fedora Linux 30
- Terminal name/version: gnome-terminal
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
:autocmd TermClose * bd!
:term
:bd!
```

### Actual behaviour

Memory is corrupted, nvim crashes immediately or soon after

### Expected behaviour

Does something reasonable

### Notes

A few minutes in gdb suggests the problem comes down to `termincal_close()` running autocmds and reentering `do_bufdel`, which is presumably not safe to reenter.

Is this related to the following test? It became flaky recently.
```
[  ERROR   ] 1 error, listed below:
[  ERROR   ] ...im/neovim/test/functional/autocmd/termclose_spec.lua @ 20: TermClose event triggers when fast-exiting terminal job stops
test/functional/helpers.lua:96: Vim(call):E900: Invalid channel id
stack traceback:
	test/functional/helpers.lua:96: in function 'request'
	test/functional/helpers.lua:254: in function 'command'
	...im/neovim/test/functional/autocmd/termclose_spec.lua:23: in function <...im/neovim/test/functional/autocmd/termclose_spec.lua:20>
 17 SKIPPED TESTS
 1 ERROR
-- Output to stderr:
CMake Error at /home/travis/build/neovim/neovim/cmake/RunTests.cmake:63 (message):
  functional tests failed with error: 1
```

--- a/src/nvim/buffer.c
+++ b/src/nvim/buffer.c
@@ -89,6 +89,7 @@
 static char *msg_loclist = N_("[Location List]");
 static char *msg_qflist = N_("[Quickfix List]");
 static char *e_auabort = N_("E855: Autocommands caused command to abort");
+static char *e_buflocked = N_("E937: Attempt to delete a buffer that is in use");
 
 // Number of times free_buffer() was called.
 static int buf_free_count = 0;
@@ -438,7 +439,7 @@ bool close_buffer(win_T *win, buf_T *buf, int action, bool abort_if_last, bool i
   // Disallow deleting the buffer when it is locked (already being closed or
   // halfway a command that relies on it). Unloading is allowed.
   if (buf->b_locked > 0 && (del_buf || wipe_buf)) {
-    emsg(_("E937: Attempt to delete a buffer that is in use"));
+    emsg(_(e_buflocked));
     return false;
   }
 
@@ -529,7 +530,9 @@ bool close_buffer(win_T *win, buf_T *buf, int action, bool abort_if_last, bool i
   }
 
   if (buf->terminal) {
+    buf->b_locked++;
     terminal_close(&buf->terminal, -1);
+    buf->b_locked--;
   }
 
   // Always remove the buffer when there is no file name.
@@ -1182,6 +1185,10 @@ int do_buffer(int action, int start, int dir, int count, int forceit)
   if (unload) {
     int forward;
     bufref_T bufref;
+    if (buf->b_locked) {
+      emsg(_(e_buflocked));
+      return FAIL;
+    }
     set_bufref(&bufref, buf);
 
     // When unloading or deleting a buffer that's already unloaded and

--- a/src/nvim/testdir/test_autocmd.vim
+++ b/src/nvim/testdir/test_autocmd.vim
@@ -169,7 +169,9 @@ func Test_autocmd_bufunload_avoiding_SEGV_01()
     exe 'autocmd BufUnload <buffer> ' . (lastbuf + 1) . 'bwipeout!'
   augroup END
 
-  call assert_fails('edit bb.txt', 'E937:')
+  " Todo: check for E937 generated first
+  " call assert_fails('edit bb.txt', 'E937:')
+  call assert_fails('edit bb.txt', 'E517:')
 
   autocmd! test_autocmd_bufunload
   augroup! test_autocmd_bufunload
@@ -540,7 +542,7 @@ func Test_three_windows()
   e Xtestje2
   sp Xtestje1
   call assert_fails('e', 'E937:')
-  call assert_equal('Xtestje2', expand('%'))
+  call assert_equal('Xtestje1', expand('%'))
 
   " Test changing buffers in a BufWipeout autocommand.  If this goes wrong
   " there are ml_line errors and/or a Crash.
@@ -563,7 +565,6 @@ func Test_three_windows()
 
   au!
   enew
-  bwipe! Xtestje1
   call delete('Xtestje1')
   call delete('Xtestje2')
   call delete('Xtestje3')

--- a/test/functional/autocmd/termxx_spec.lua
+++ b/test/functional/autocmd/termxx_spec.lua
@@ -7,6 +7,8 @@ local eval, eq, neq, retry =
   helpers.eval, helpers.eq, helpers.neq, helpers.retry
 local ok = helpers.ok
 local feed = helpers.feed
+local pcall_err = helpers.pcall_err
+local assert_alive = helpers.assert_alive
 local iswin = helpers.iswin
 
 describe('autocmd TermClose', function()
@@ -14,6 +16,24 @@ describe('autocmd TermClose', function()
     clear()
     nvim('set_option', 'shell', testprg('shell-test'))
     command('set shellcmdflag=EXE shellredir= shellpipe= shellquote= shellxquote=')
+  end)
+
+
+  local function test_termclose_delete_own_buf()
+    command('autocmd TermClose * bdelete!')
+    command('terminal')
+    eq('Vim(bdelete):E937: Attempt to delete a buffer that is in use', pcall_err(command, 'bdelete!'))
+    assert_alive()
+  end
+
+  -- TODO: fixed after merging patches for `can_unload_buffer`?
+  pending('TermClose deleting its own buffer, altbuf = buffer 1 #10386', function()
+    test_termclose_delete_own_buf()
+  end)
+
+  it('TermClose deleting its own buffer, altbuf NOT buffer 1 #10386', function()
+    command('edit foo1')
+    test_termclose_delete_own_buf()
   end)
 
   it('triggers when fast-exiting terminal job stops', function()

